### ISSUE TITLE
Why does the shopify_app use the session instead of hmac?

### ISSUE THREAD

**Issue by atomgiant** [Shopify/shopify_app#204]
The shopify_app is using the session to verify embedded app requests:

https://github.com/Shopify/shopify_app/blob/33f308a454be94e887bba80e39b6e4f8b76caae1/lib/shopify_app/sessions_controller.rb#L19

For an incoming request, I would expect the hmac + shop params could be used to verify the request is allowed for an embedded Shopify app.

The hmac approach has a couple pros I believe make it a better choice than the session approach:

1) There is no need to use the HTTP session.  Each request contains everything you need via the params so it is stateless.  (The app would still need to store the shop and token info in a database or other persistent storage).

2) If the store admin's browser session is cleared, or additional web browsers are used, the OAuth flow can be skipped since everything needed would be in the url params.

I believe the current session-based approach will cause the OAuth flow to happen every time the session expires or doesn't exist for a given store.

Thanks for your help.

- Tom

**Comment by kevinhughes27** [comment]
You're right the current approach does result in the OAuth flow being re-done when the session expires or doesn't exist (like a new browser). If you would like to submit a PR to use the shop param + hmac we can have a look and discuss further. It would be nice if it was configurable for now leaving the current session based method in tact.

**Comment by atomgiant** [comment]
Hey @kevinhughes27 - Thanks for your quick reply. 

I mainly wanted to make sure I wasn't missing something.  So it sounds like the hmac + shop is feasible.

I think the hmac + shop param approach may be tough to make work in parallel with the current session approach without doing a something like trying one and falling back to the other.  I appreciate the need to have continuity for this project though.

I think in the short-term I am going to end up using my own hmac + shop approach rather than a PR so you can close this "issue".   If time permits I will at least document my approach and possibly contribute back a README update in the future.

Thanks!

**Comment by kevinhughes27** [comment]
Makes sense. It would be great if you share your solution either way! I'd be interested in seeing how it works out. I'll leave the issue open for now.

**Comment by EiNSTeiN-** [comment]
An additional problem with using the signed parameters from Shopify is that they are not always available (if the user clicks on a link inside your app, the next request won't have them) so it is not 100% stateless, you still need to maintain a session (unless your app has a single page...). For non-embedded app these parameters are also available only on the very first redirect from Shopify, not on subsequent requests. I think the current approach fits the majority of apps better. I'm going to close this since I don't think we have any plans to change the current approach.

**Comment by sheppyh** [comment]
Could my question here be related to this?
#574 
@atomgiant I wonder how you got on with your own HMAC implementation.

Cheers

**Comment by atomgiant** [comment]
@sheppyh I'm not sure about Android but I think it should work the same.  shopify_app does require sessions so I'd imagine the Android app is using sessions.  

For completeness, here is a follow up of what I found with trying to switch to HMAC-only for auth.  I did implement my own HMAC to auto login, but I ended up having to switch back to sessions for two reasons:

1. Using only the HMAC required me to recompute the HMAC for all internal links for my app.
2. The shopify app team runs a test against apps where they uninstall and then try to hit an existing app link.  The HMAC only approach was (incorrectly) allowing the link to work because it bypassed the OAuth flow.

Sorry I can't really help with your Android issue since I don't have any experience there.
